include/rmm/mr/device/cuda_async_memory_resource.hpp

```diff
@@ -0,0 +1,154 @@
+
+  ~cuda_async_memory_resource()
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    cudaDeviceSynchronize();
+    int device{0};
+    RMM_ASSERT_CUDA_SUCCESS(cudaGetDevice(&device));
+    cudaMemPool_t pool;
+    RMM_ASSERT_CUDA_SUCCESS(cudaDeviceGetDefaultMemPool(&pool, device));
+    RMM_ASSERT_CUDA_SUCCESS(cudaMemPoolTrimTo(pool, 0));
+#endif
+  }
+
+  cuda_async_memory_resource(cuda_async_memory_resource const&) = default;
+  cuda_async_memory_resource(cuda_async_memory_resource&&)      = default;
+  cuda_async_memory_resource& operator=(cuda_async_memory_resource const&) = default;
+  cuda_async_memory_resource& operator=(cuda_async_memory_resource&&) = default;
+
+  /**
+   * @brief Query whether the resource supports use of non-null CUDA streams for
+   * allocation/deallocation. `cuda_memory_resource` does not support streams.
+   *
+   * @returns bool true
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return true
+   */
+  bool supports_get_mem_info() const noexcept override { return true; }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using cudaMalloc.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, rmm::cuda_stream_view stream) override
+  {
+    void* p{nullptr};
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    if (bytes > 0) { RMM_CUDA_TRY(cudaMallocAsync(&p, bytes, stream.value()), rmm::bad_alloc); }
+#endif
+    return p;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   */
+  void do_deallocate(void* p, std::size_t, rmm::cuda_stream_view stream) override
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    if (p != nullptr) { RMM_ASSERT_CUDA_SUCCESS(cudaFreeAsync(p, stream.value())); }
+#endif
+  }
+
+  /**
+   * @brief Compare this resource to another.
+   *
+   * @throws Nothing.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept override
+  {
+    return dynamic_cast<cuda_async_memory_resource const*>(&other) != nullptr;
+  }
+
+  /**
+   * @brief Get free and available memory for memory resource
+   *
+   * @throws `rmm::cuda_error` if unable to retrieve memory info.
+   *
+   * @return std::pair contaiing free_size and total_size of memory
+   */
+  std::pair<size_t, size_t> do_get_mem_info(rmm::cuda_stream_view) const override
+  {
+    std::size_t free_size;
+    std::size_t total_size;
+    RMM_CUDA_TRY(cudaMemGetInfo(&free_size, &total_size));
```

Comment:
Is it RMM's expectation that there's only one memory resource in use at a time, so `cudaMemGetInfo`'s output can be considered close enough, or that it's OK to just have a rough estimate?

Reply:
Ah, this function is TODO. Thanks for the reminder.

Reply:
I thought maybe the new CUDA pool functionality would have a way to query free memory for a pool, but there doesn't appear to be. So I think this function should be removed from this MR.

Reply:
Disabled.